File: Data Structure ZJU/FileTransfer.cpp
#include<cstdio>

const int maxn = 10010;
typedef int ElementType;
typedef int setName;
typedef int setType[maxn];

void Initialization(setType s,int n);
int Find(setType s,ElementType x);
void Input_connection(setType s);
void Check_connection(setType s);
void Check_network(setType s,int n);

int main(){
    int n;
    char in;
    setType s;
    scanf("%d\n",&n);    
    Initialization(s,n);
    do{
        scanf("%c",&in);
        switch(in){
            case 'I':Input_connection(s);break;
            case 'C':Check_connection(s);break;
            case 'S':Check_network(s,n);break;
        }
    }while(in != 'S');
    return 0;
}

void Input_connection(setType s){
    ElementType u,v;
    scanf("%d%d",&u,&v);
    setName root1,root2;
    root1 = Find(s,u-1);
    root2 = Find(s,v-1);
    if(root1 != root2) s[root1] = root2;
}

void Initialization(setType s,int n){
    for(int i = 0; i < n; i++){
        s[i] = -1;
    }
}

void Check_connection(setType s){
    ElementType u,v;
    scanf("%d%d",&u,&v);
    setName root1 = Find(s,u-1);
    setName root2 = Find(s,v-1);
    if(root1 != root2) printf("no\n");
    else printf("yes\n");
}

void Check_network(setType s,int n){
    int cnt = 0;
    for(int i = 0; i < n; i++){
        if(s[i] < 0) cnt++;
    }
    if(cnt == 1) printf("The network is connected.\n");
    else printf("There are %d components.",cnt);
}

int Find(setType s,ElementType x){
    if(s[x] < 0) return x;
    else return s[x] = Find(s,s[x]);
}

File: Data Structure ZJU/FileTransfer.py
def FindRoot(c):
    index = c
    while sList[index] > 0:
        index = sList[index]
    return index

def check(c1, c2):
    r1 = FindRoot(c1)
    r2 = FindRoot(c2)
    if r1 == r2:
        print('yes')
    else:
        print('no')

def unite(c1, c2):
    r1 = FindRoot(c1)
    r2 = FindRoot(c2)
    if ( abs(sList[r1]) > abs(sList[r2]) ):
        sList[r1] -= 1
        sList[r2] = r1
    else:
        sList[r2] -= 1
        sList[r1] = r2

N = int(input())

sList=[0]+[-1]*N

while True:
    inputLine = input().split()
    cmd = inputLine[0]
    if cmd == 'S':
        rootList = [x for x in sList if x < 0]
        k = len(rootList)
        if k != 1:
            print('There are {} components.'.format(k))
        else:
            print('The network is connected.')
        break
    else:
        c1 = int(inputLine[1])
        c2 = int(inputLine[2])
        if cmd == 'I':
            unite(c1, c2)
        if cmd == 'C':
            check(c1, c2)
